src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/ILaunchSettingsProvider3.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Composition;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Debug
+{
+    /// <summary>
+    ///     Provides an implementation of <see cref="ILaunchSettingsProvider"/> with
+    ///     additional methods for atomically updating a launch profile or global setting.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    ///     The existing <see cref="ILaunchSettingsProvider.AddOrUpdateProfileAsync"/> method
+    ///     does not work well in situations where multiple update operations may run
+    ///     concurrently because the retrieval of the current set of values is separated from
+    ///     the later update. If operations A and B each retrieve the current profile, make
+    ///     their changes, and then apply them there is a very good chance that A's changes
+    ///     will overwrite B's or the other way around. The new method introduced here makes
+    ///     retrieving and updating a profile an single operation.
+    /// </para>
+    /// <para>
+    ///     The <see cref="ILaunchSettingsProvider.AddOrUpdateGlobalSettingAsync"/> method has
+    ///     similar problems.
+    /// </para>
+    /// </remarks>
+    [ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.Private, Cardinality = ImportCardinality.ExactlyOne)]
+    public interface ILaunchSettingsProvider3 : ILaunchSettingsProvider2
+    {
+        /// <summary>
+        ///     Supports the retrieval and update of a given launch profile as a single
+        ///     operation. When <paramref name="updateAction"/> is called it is given an
+        ///     <see cref="IWritableLaunchProfile"/> with the current state of the
+        ///     <paramref name="profileName"/> profile. It will update the profile as
+        ///     appropriate and when it is done the profile is applied to the settings.
+        /// </summary>
+        Task UpdateProfileAsync(string profileName, Action<IWritableLaunchProfile> updateAction);
+
+        /// <summary>
+        ///     Supports the retrieval and update of a given global setting as a single
+        ///     operation. When <paramref name="updateFunction"/> is called it is given the
+        ///     current value of the <paramref name="settingName"/> global setting (which may be
+        ///     <c>null</c> if no such setting exists and returns the updated object (or
+        ///     <c>null</c> if the setting is to be removed).
```

Comment:
Using `<cref langword="null"/>` provides a better cross-language experience. For example, `Nothing` is shown in VB.NET.

Reply:
I always forget about `langword="null"`.